### ISSUE TITLE
Subpar font hinting on low-DPI displays

### ISSUE THREAD

**Issue by LunarLambda** [pop-os/cosmic-term#363]
The default hinting used by cosmic-term/cosmic-text looks doesn't look as good as most other apps/terminals, even at larger font sizes (15pt/20px in my case). Edges are overly sharp and there's no subpixel rendering. It's particularly noticeable on `o`s, `e`s, as well as on brightly colored text, where stroke widths can end up looking slightly uneven.

Example: Top: `cosmic-term`, Bottom: `alacritty`
![image](https://github.com/user-attachments/assets/6b1fb979-b2a9-4f7f-9f3a-b9e3542e6f8e)

Zoomed in:
![image](https://github.com/user-attachments/assets/fb5f00a9-ad59-42a1-a72f-6f08262e3ba5)

`wezterm` with subpixel rendering (`freetype_render_target = 'HorizontalLcd'`) gets pixel-for-pixel identical looking results to `alacritty` (minus differences in colorschemes), as does `gnome-terminal`.

**Comment by charliechensz** [comment]
Confirming this issue is still present in 1.5.0, with a concrete data point: **cosmic-term 1.5.0 changed its render stack** (1.3.0 → 1.5.0: from system fontconfig to an embedded one, using the skrifa/swash shaping stack), and font rendering quality visibly regressed — grainy/fuzzy glyphs at typical sizes, notably lacking subpixel antialiasing, while kitty on the same system (freetype + subpixel AA) renders cleanly.

Environment: Pop!_OS 24.04, cosmic-term 1.5.0, NVIDIA RTX 4060, typical desktop monitor (~96 DPI).

Observations:
- Low-DPI displays are where this hurts most — exactly the "overly sharp edges, no subpixel rendering" described in the screenshots here.
- The render-stack change in 1.5 did fix a separate issue (glyph drift in Neovim), so it is not a pure regression — but font appearance is a real downgrade.
- Workaround for now: I switched to kitty. But cosmic-term is the default terminal, so it would be great to see this fixed.

Suggestions:
1. Enable subpixel antialiasing (or at least better hinting) — critical on ≤96 DPI displays.
2. Check whether the embedded fontconfig config disables hinting/subpixel by default.
3. Expose an AA mode + hinting strength option in advanced font settings.

Happy to provide side-by-side screenshots (cosmic-term vs kitty) if helpful.

**Comment by charliechensz** [comment]
## Additional data point: same version, night-and-day difference explained by PPI — and a failed self-compile attempt that pinpoints the blocker

I hit this same issue and did a fairly thorough investigation. Sharing the results in case it helps narrow the fix.

### Two-machine controlled comparison

Same cosmic-term version (1.8.0), same config, only the panel differs:

| Machine | Panel | PPI | Result |
|---|---|---|---|
| A (ms760) | 31.5″ 4K LCD | ~140 | visible fringing/jagged CJK & Latin, esp. stroke-heavy text |
| B (xm16) | 15.7″ 3.2K OLED | ~259 | looks fine |

Same font, same size, same EDID subpixel declaration (both BGR). The differentiator is purely **grayscale AA at low PPI**: cosmic-text hardcodes `Format::Alpha` (grayscale) — no LCD subpixel rendering at all — which halves effective horizontal resolution. At 259 PPI that's imperceptible; at 140 PPI it's obvious. Font/size changes don't help (confirmed by controlled A/B).

### Self-compile attempt: patch works at the library level, fails end-to-end

I applied cosmic-text PR #532 (`CacheKeyFlags::SUBPIXEL_RGB/BGR` + LCD subpixel mask rasterization), wired `SUBPIXEL_BGR` into cosmic-term's two `Attrs::new()` sites, and built 1.8.0.

- Library-level: the patched cosmic-text correctly emits `SwashContent::SubpixelMask` with the expected BGR stripe order (verified pixel-level).
- End-to-end: **rendering became blurry** — because glyphon's `text_render.rs` explicitly does:

```rust
SwashContent::SubpixelMask => {
    // Not implemented yet, but don't panic if this happens.
    ContentType::Mask
}
```

So the consumer pipeline doesn't know how to composite a subpixel mask yet. A production fix needs **both** sides: cosmic-text producing the mask (PR #532) **and** glyphon learning to composite `SubpixelMask` (RGBA content + proper subpixel blending) instead of falling back to `Mask`.

### What this means for #363

The missing piece is essentially: (a) land a subpixel-mask producer in cosmic-text, and (b) teach glyphon to consume it. Until both exist, low-DPI users (≤~150 PPI) will keep seeing the soft/jagged rendering described here. Happy to test a combined change on the 140 PPI panel if someone picks this up.

**Comment by charliechensz** [comment]
## PSA: paths that do NOT help (verified by controlled testing) — so others don't waste time

Following up on my earlier comment — I also tested the common workarounds people try for this issue, so you don't have to. All of the following were **verified ineffective** on cosmic-term 1.8.0:

### 1. `fontconfig` rendering settings (rgba / hintstyle) — dead end
Editing `~/.config/fontconfig/fonts.conf` to set `<rgba>rgb</rgba>` / `<hintstyle>hintfull</hintstyle>` has **no effect**. cosmic-text does not read fontconfig rendering parameters — its fontconfig integration is only used for *font discovery* (via `fontdb`). The subpixel/hinting/AA behavior is hardcoded in the rasterizer (`Format::Alpha` grayscale in `swash.rs`). Verified in source.

### 2. `wl_output` subpixel order — dead end
The compositor advertises the panel's subpixel order (our monitor correctly reports BGR via EDID), but cosmic-term/cosmic-text **never reads it**. There is no code path that consults the output's subpixel geometry. Also verified in source.

### 3. Font family / size changes — no effect on sharpness
Tested across DejaVu Sans Mono, Noto Sans Mono, Noto Sans Mono CJK SC, sizes 14→17pt (21% larger, clearly visible). Glyph sharpness was **unchanged** — only size changed. The fringing is a property of the grayscale rasterizer at low PPI, not the font or size. Verified by controlled A/B on two machines (see my previous comment: 140 PPI vs 259 PPI, identical version/config — the difference is purely PPI).

### 4. Verifying your panel's real subpixel order (if you want to report it)
Read the EDID directly rather than trusting drivers/labels:
```
od -An -tu1 -j18 -N1 /sys/class/drm/card0-DP-3/edid
```
- `0x01` = RGB, `0x02` = BGR (bit0-1 of byte 18). Earlier I misread a "panel-specific" value and chased a wrong subpixel-phasing theory — measuring the actual EDID is the reliable path.

### Bottom line
None of the user-facing configuration can improve this — it's a rasterizer limitation (grayscale AA only) that becomes visible below ~150 PPI. The fix needs to land in code: a subpixel-mask producer in cosmic-text (PR #532) **plus** a consumer in glyphon (currently `SubpixelMask => ContentType::Mask` / "not implemented yet"). Details in my previous comment.
